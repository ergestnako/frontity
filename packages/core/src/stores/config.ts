import { Namespace } from "@frontity/types/namespace";
import { namespaced } from "overmind/config";
import { Settings } from "./settings/types";

export default ({
  namespaces,
  settings
}: {
  namespaces: { [key: string]: Namespace };
  settings: Settings;
}) =>
  namespaced(
    Object.entries(namespaces).reduce(
      (config, [namespace, module]) => {
        if (module.Store) config[namespace] = module.Store;
        return config;
      },
      { settings }
    )
  );
