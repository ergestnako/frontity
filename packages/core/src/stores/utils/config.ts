import { Namespace } from "@frontity/types/namespace";
import { namespaced } from "overmind/config";

export default ({ namespaces }: { namespaces: { [key: string]: Namespace } }) =>
  namespaced(
    Object.entries(namespaces).reduce((config, [namespace, module]) => {
      if (module.Store) config[namespace] = module.Store;
      return config;
    }, {})
  );
