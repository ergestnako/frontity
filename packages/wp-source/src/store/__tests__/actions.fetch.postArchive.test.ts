import { createOvermindMock } from "overmind";
import { namespaced } from "overmind/config";
import { Response, Headers } from "cross-fetch";
import { state, actions, effects } from "..";

import patterns from "./mocks/patterns";
import postArchiveJson from "./mocks/postArchive.json";

describe("fetch", () => {
  test("get post archive", async () => {
    const { resolver } = effects;
    patterns.forEach(({ pattern, handler }) => {
      resolver.add(pattern, handler);
    });
    const api = {
      get: jest.fn(() =>
        Promise.resolve(
          new Response(JSON.stringify(postArchiveJson), {
            headers: new Headers({
              "X-WP-Total": "47",
              "X-WP-TotalPages": "5"
            })
          })
        )
      )
    };
    const config = namespaced({ source: { actions, state, effects } });
    const store = createOvermindMock(config, {
      source: { resolver, api }
    });
    await store.actions.source.fetch({ name: "/" });
    expect(api.get.mock.calls).toMatchSnapshot();
    expect(api.get).toBeCalledTimes(1);
    expect(store.state.source.dataMap).toMatchSnapshot();
  });

  test("post archive twice gets data once", async () => {
    const { resolver } = effects;
    patterns.forEach(({ pattern, handler }) => {
      resolver.add(pattern, handler);
    });
    const api = {
      get: jest.fn(() =>
        Promise.resolve(
          new Response(JSON.stringify(postArchiveJson), {
            headers: new Headers({
              "X-WP-Total": "47",
              "X-WP-TotalPages": "5"
            })
          })
        )
      )
    };
    const config = namespaced({ source: { actions, state, effects } });
    const store = createOvermindMock(config, {
      source: { resolver, api }
    });
    await store.actions.source.fetch({ name: "/" });
    await store.actions.source.fetch({ name: "/" });
    expect(api.get).toBeCalledTimes(1);
  });
});