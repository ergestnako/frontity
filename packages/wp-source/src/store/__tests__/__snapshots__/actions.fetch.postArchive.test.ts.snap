// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch get post archive 1`] = `
Array [
  Array [
    Object {
      "endpoint": "posts",
      "params": Object {
        "_embed": true,
        "page": 1,
        "search": undefined,
      },
    },
  ],
]
`;

exports[`fetch get post archive 2`] = `
Object {
  "/": Object {
    "isArchive": true,
    "isFetching": false,
    "isHome": true,
    "isPostArchive": true,
    "isPostTypeArchive": true,
    "isReady": true,
    "pages": Array [
      Array [
        Object {
          "id": 1801,
          "link": "https://test.frontity.io/2018/block-image/",
          "type": "post",
        },
        Object {
          "id": 1798,
          "link": "https://test.frontity.io/2018/block-button/",
          "type": "post",
        },
        Object {
          "id": 1797,
          "link": "https://test.frontity.io/2018/block-cover/",
          "type": "post",
        },
        Object {
          "id": 1796,
          "link": "https://test.frontity.io/2018/column-blocks/",
          "type": "post",
        },
        Object {
          "id": 1800,
          "link": "https://test.frontity.io/2018/block-gallery/",
          "type": "post",
        },
        Object {
          "id": 1799,
          "link": "https://test.frontity.io/2018/block-quotes/",
          "type": "post",
        },
        Object {
          "id": 1786,
          "link": "https://test.frontity.io/2018/block-category-common/",
          "type": "post",
        },
        Object {
          "id": 1789,
          "link": "https://test.frontity.io/2018/blocks-embeds/",
          "type": "post",
        },
        Object {
          "id": 1795,
          "link": "https://test.frontity.io/2018/blocks-widgets/",
          "type": "post",
        },
        Object {
          "id": 1788,
          "link": "https://test.frontity.io/2018/blocks-layout-elements/",
          "type": "post",
        },
      ],
    ],
    "total": 47,
    "totalPages": 5,
    "type": "post",
  },
  "/2018/block-button/": Object {
    "id": 1798,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/2018/block-category-common/": Object {
    "id": 1786,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/2018/block-cover/": Object {
    "id": 1797,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/2018/block-gallery/": Object {
    "id": 1800,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/2018/block-image/": Object {
    "id": 1801,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/2018/block-quotes/": Object {
    "id": 1799,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/2018/blocks-embeds/": Object {
    "id": 1789,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/2018/blocks-layout-elements/": Object {
    "id": 1788,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/2018/blocks-widgets/": Object {
    "id": 1795,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/2018/column-blocks/": Object {
    "id": 1796,
    "isFetching": false,
    "isPost": true,
    "isPostType": true,
    "isReady": true,
    "type": "post",
  },
  "/author/themereviewteam/": Object {
    "id": 6,
    "isArchive": true,
    "isAuthor": true,
    "isFetching": false,
    "isReady": true,
    "pages": Array [],
  },
  "/category/block/": Object {
    "id": 66,
    "isArchive": true,
    "isCategory": true,
    "isFetching": false,
    "isReady": true,
    "isTaxonomy": true,
    "pages": Array [],
    "taxonomy": "category",
  },
  "/category/media-2/": Object {
    "id": 92,
    "isArchive": true,
    "isCategory": true,
    "isFetching": false,
    "isReady": true,
    "isTaxonomy": true,
    "pages": Array [],
    "taxonomy": "category",
  },
  "/tag/columns/": Object {
    "id": 253,
    "isArchive": true,
    "isFetching": false,
    "isReady": true,
    "isTag": true,
    "isTaxonomy": true,
    "pages": Array [],
    "taxonomy": "tag",
  },
  "/tag/embeds-2/": Object {
    "id": 150,
    "isArchive": true,
    "isFetching": false,
    "isReady": true,
    "isTag": true,
    "isTaxonomy": true,
    "pages": Array [],
    "taxonomy": "tag",
  },
  "/tag/gallery/": Object {
    "id": 162,
    "isArchive": true,
    "isFetching": false,
    "isReady": true,
    "isTag": true,
    "isTaxonomy": true,
    "pages": Array [],
    "taxonomy": "tag",
  },
  "/tag/image/": Object {
    "id": 169,
    "isArchive": true,
    "isFetching": false,
    "isReady": true,
    "isTag": true,
    "isTaxonomy": true,
    "pages": Array [],
    "taxonomy": "tag",
  },
  "/tag/video/": Object {
    "id": 233,
    "isArchive": true,
    "isFetching": false,
    "isReady": true,
    "isTag": true,
    "isTaxonomy": true,
    "pages": Array [],
    "taxonomy": "tag",
  },
}
`;
